### ISSUE TITLE
JIT compiling this specific function with grad-enabled tensors makes it always produce grad-enabled tensors

### ISSUE THREAD

**Issue by martenlienen** [pytorch/pytorch#83378]
### 🐛 Describe the bug

I into this weird JIT bug a while back and reduced it into a minimal example today. Consider the expression `a + (b / (-b))[:]` where `a` and `b` are tensors of appropriate shape. If we put this expression into a function and JIT compile it, it will work with the following caveat: If the JIT compilation is triggered by a call where `a.requires_grad`, the compiled function will always return a tensor that `.requires_grad` and has a backward graph attached, regardless of any of the inputs requiring gradients or not. Triggering the JIT compilation with tensors that do not require gradients will produce a function that behaves as expected. Curiously, the behavior depends upon the formulation of the operations on `b`, the tensor that *does not even require a gradient*. If we remove the minus, for example, `a + (b / b)[:]` or move the minus into a different position `a + (-(b / b)[:])`, JIT compilation works as expected.

The code below reproduces this behavior. It says that a function "fails", if JIT compiling it with grad-requiring tensors makes it always return grad-requiring tensors, and it "works" otherwise.

```python
import torch

def check_no_grad(f):
    f_ = torch.jit.script(f)

    y = torch.ones(1)
    t = torch.ones(1)
    assert not t.requires_grad

    y.requires_grad_()
    assert y.requires_grad
    assert f_(y, t).requires_grad

    y.requires_grad_(False)
    assert not y.requires_grad
    if f_(y, t).requires_grad:
        print(f.__name__, "fails")
    else:
        print(f.__name__, "works")

def works_1(a, b):
    x = b / (-b)
    return a + x

def works_2(a, b):
    x = (b / b)[:]
    return a + x

def works_3(a, b):
    x = -(b / b)[:]
    return a + x

def fail_1(a, b):
    x = (b / (-b))[:]
    return a + x

check_no_grad(works_1)
check_no_grad(works_2)
check_no_grad(works_3)
check_no_grad(fail_1)
```

```
$ python bug.py
works_1 works
works_2 works
works_3 works
fail_1 fails
```

### Versions

PyTorch version: 1.12.0+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 12.1.1 20220730
Clang version: 14.0.6
CMake version: version 3.23.3
Libc version: glibc-2.36

Python version: 3.10.5 (main, Jul 15 2022, 14:39:13) [GCC 12.1.0] (64-bit runtime)
Python platform: Linux-5.15.59-1-MANJARO-x86_64-with-glibc2.36
Is CUDA available: False
CUDA runtime version: 11.7.99
GPU models and configuration: GPU 0: NVIDIA T1200 Laptop GPU
Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] functorch==0.2.0
[pip3] numpy==1.23.1
[pip3] pytorch-lightning==1.6.4
[pip3] torch==1.12.0+cu116
[pip3] torchmetrics==0.9.2
[pip3] torchode==0.1.0
[pip3] torchtyping==0.1.4
[conda] Could not collect

**Comment by davidberard98** [comment]
@martenlienen I believe this has been fixed in #79498 and #78392, and I was able to confirm that on master this test passes.

But - thanks for the report and repro! I had a pretty hard time finding this issue when I saw it a few months ago, so I appreciate the effort you took to isolate the issue 😄
